Framework/Algorithms/src/CreateMonteCarloWorkspace.cpp

```diff
@@ -0,0 +1,163 @@
+                  "Name of output workspace.");
+}
+
+//----------------------------------------------------------------------------------------------
+
+Mantid::HistogramData::HistogramY CreateMonteCarloWorkspace::fillHistogramWithRandomData(const std::vector<double> &cdf,
+                                                                                         int numIterations,
+                                                                                         int seedInput,
+                                                                                         API::Progress &progress) {
+
+  Mantid::HistogramData::HistogramY outputY(cdf.size(), 0.0);
+  std::mt19937 gen(seedInput);
+  std::uniform_real_distribution<> dis(0.0, 1.0);
+
+  int progressInterval = std::max(1, numIterations / 100); // Update progress every 1%
+
+  for (int i = 0; i < numIterations; ++i) {
+    double randomNum = dis(gen);
+    auto it = std::lower_bound(cdf.begin(), cdf.end(), randomNum);
+    size_t index = std::distance(cdf.begin(), it);
+
+    if (index < outputY.size()) {
+      outputY[index] += 1.0;
+    }
+
+    if (i % progressInterval == 0) {
+      progress.report("Generating random data...");
+    }
+  }
+  return outputY;
+}
+
+/**
+ *  Compute a normalized CDF [0..1] from the given histogram data.
+ */
+std::vector<double> CreateMonteCarloWorkspace::computeNormalizedCDF(const Mantid::HistogramData::HistogramY &yData) {
+  std::vector<double> cdf(yData.size());
+  std::partial_sum(yData.begin(), yData.end(), cdf.begin());
+  double totalCounts = cdf.back();
+
+  if (totalCounts > 0.0) {
+    // Normalize the CDF
+    std::transform(cdf.begin(), cdf.end(), cdf.begin(), [totalCounts](double val) { return val / totalCounts; });
+  } else {
+    g_log.warning("Total counts are zero; normalization skipped.");
+  }
+  return cdf;
+}
+
+/**
+ *  Determine how many iterations to use for MC sampling.
+ *  If userMCEvents > 0, use that directly; otherwise use the integral of the input data.
+ */
+int CreateMonteCarloWorkspace::integrateYData(const Mantid::HistogramData::HistogramY &yData) {
+  double total_counts = std::accumulate(yData.begin(), yData.end(), 0.0);
+  int iterations = static_cast<int>(std::round(total_counts));
+
+  if (iterations == 0) {
+    g_log.warning("Total counts in the input workspace round to 0. No Monte Carlo events will be generated.");
+  }
+  return iterations;
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void CreateMonteCarloWorkspace::exec() {
+  MatrixWorkspace_sptr inputWs = getProperty("InputWorkspace");
+  int seedInput = getProperty("Seed");
+  int userMCEvents = getProperty("MonteCarloEvents");
+
+  const auto &originalYData = inputWs->y(0); // Counts in each bin
+
+  int numIterations = (userMCEvents > 0) ? userMCEvents : integrateYData(originalYData);
+
+  API::Progress progress(this, 0.0, 1.0, 2);
```

Comment:
```suggestion
  API::Progress progress(this, 0.0, 1.0, 101);
```
Setting the progress reporter to have 101 steps (one for the cdf creation, and 100 for the percentages of the MC generation steps), then it starts to work as expected. I tested this by setting MonteCarloEvents=10000 to make it take 10 seconds to generate.

Reply:
Alternatively (maybe a neater solution because you don't need to pass the reporter through to the function?), you can just create a new progress reporter inside the generating function and give it 100 steps there, that seems to work fine when I test it.